### PR TITLE
Posix shell

### LINKED ISSUE #16055: unix gradle wrapper fails when path contains $ character
### Expected Behavior
gradle wrapper works regardless of the name of the current directory or its ancestors.

### Current Behavior

When the project directory or any of its ancestors contain a `$` character, `gradlew` fails with

```
Error: Could not find or load main class org.gradle.wrapper.GradleWrapperMain
Caused by: java.lang.ClassNotFoundException: org.gradle.wrapper.GradleWrapperMain
```

### Context

I was debugging problems in our build that occur when there is a space in the name (e.g. home directories where the username is multiple words). For extra spice, I added some other characters to my directory names too.

`%` and `👿` cause problems that might be at a deeper level, but the `$` is specific to the gradlew shell script.

### Steps to Reproduce 

```
mkdir 'foo$bar-baz' && cd 'foo$bar-baz'
gradle init
./gradlew
```

### Your Environment

Ubuntu 20.10 here, but I suspect this is widely reproducible.

### PR DESCRIPTION
**Pull request by kurahaupo**
Fixes #16055, #17215, [gradle-private#3386](/gradle/gradle-private/issues/3386) and possibly others TBA

### Context
Mostly adjustments to the `gradlew` script:
* Allow `gradlew` to pass shellcheck.
* Restore "any POSIX shell" as prerequisite (not just Bash, which was recently made a requirement).
* Address significant technical debt.

I'm approaching this from the perspective of an experienced shell user who encountered the `gradlew` script inside another project, and I was frankly horrified when I opened the lid.

This has turned out to be a lot steeper learning curve and a lot more work than I expected, so I would like early feedback on whether any of it is likely to be accepted by the project.

### Contributor Checklist
- [X] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [X] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [X] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [X] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team

Remaining answers are "not yet".
* no new functionality, so it's unclear what tests I should add
* I don't yet have an environment where I can run the test suite.

- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by kurahaupo** [comment]
I have a few questions about `gradlew` and I'm not really clear where I should ask them.

Firstly, is any of this worth my time?

Commit 4cca85e was to ensure that old pre-POSIX shells were still supported, but commit ff09404 jumped to the other extreme and now it doesn't support POSIX shells, just Bash. This PR reverses that: any POSIX shell should suffice. (Where `/bin/sh` is a pre-POSIX shell, gradlew can still be run but needs to be prefaced by the name of a POSIX shell; `ksh` suffices on Solaris.)

Secondly, there are a bunch of template variables `subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt`, which as far as I can see are fixed boilerplate. Is there any reason for them to remain, rather than use their assigned values directly?

    ${appNameSystemProperty}    org.gradle.appname
    ${applicationName}          Gradle
    ${mainClassName}            org.gradle.wrapper.GradleWrapperMain
    ${optsEnvironmentVar}       GRADLE_OPTS

I assume that `${defaultJvmOpts}` will vary for the system gradlew is built on?

Thirdly, I note that `GRADLE_OPTS` was hard-coded in some places but `${optsEnvironmentVar}` was used elsewhere; was that combination intentional?

**Comment by kurahaupo** [comment]
You may notice that I replaced `#!/usr/bin/env sh` with `#!/bin/sh`.

If the previous concerns about executable code in environment variables is valid, then unsafe entries in `PATH` in particular are also an issue. Removing `/usr/bin/env` addresses this.

In any case, `/bin/sh` and `/usr/bin/env` have the same near-certain guarantees from POSIX, and in practice `/bin/sh` has fewer unsupported systems. So we might as well close the hole.

**Comment by ljacomet** [comment]
@big-guy Can you take a look here given the work around switching to bash?

File: subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt [hunk 1]
```diff
@@ -17,37 +17,97 @@
 #
 
 ##############################################################################
-##
-##  ${applicationName} start up script for UN*X
```
Comment (big-guy) [inline subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt:21]: This is used to also generate start scripts for Java applications that apply the `application` plugin, so the substitutions need to stay.

File: subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt [hunk 2]
```diff
@@ -17,37 +17,97 @@
...
+#       various environments, including this one. You can find it at
+#       https://github.com/gradle/gradle
+#
+#   (4) Because we're targeting any POSIX shell, we progressively modify "$@",
+#       rather than try and built up all the args just before invoking Java.
+#       We even pass shellcheck.
+#
+#   (5) This script (gradlew) is generated from
+#       subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt
+#       in the gradle project, so submitting patches will be tedious.
```
Comment (big-guy) [inline subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt:55]: As I said above, this script is used for things beyond `gradlew`, so it probably doesn't make sense to include these details that may only apply to Gradle or builds.

It's sort of useful to know that this is a generated file and it was generated by Gradle

File: subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt [hunk 3]
```diff
@@ -17,37 +17,97 @@
...
+#       ${mainClassName}            org.gradle.wrapper.GradleWrapperMain
+#       ${optsEnvironmentVar}       GRADLE_OPTS
+#
+#    (b) some info about your build settings:
+#
+#       ${defaultJvmOpts}           varies, but here it's:
+#                                   -Dfile.encoding=UTF-8 "-Xmx64m" "-Xms64m"
+%>
+#
 ##############################################################################
```
Comment (big-guy) [inline subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt:81]: To try this locally, I ended up just deleting this entire comment block. I think the unescaped `$` in here breaks the generator.

**Review by big-guy — CHANGES_REQUESTED** [review]
> I have a few questions about `gradlew` and I'm not really clear where I should ask them.
> 
> Firstly, is any of this worth my time?
> 

Yes, this is greatly appreciated. Both Vaidotas and I knew this was a mess, but we didn't have enough experience to make it better.

> Secondly, there are a bunch of template variables `subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt`, which as far as I can see are fixed boilerplate. Is there any reason for them to remain, rather than use their assigned values directly?

Yes, this needs to remain. The same template is used to generate start scripts for any Java project that uses the `application` plugin, so the value of these variables can be different and non-Gradle specific.

> I assume that `${defaultJvmOpts}` will vary for the system gradlew is built on?

No, they're the same for all `gradlew` and potentially different for non-Gradle projects.

> Thirdly, I note that `GRADLE_OPTS` was hard-coded in some places but `${optsEnvironmentVar}` was used elsewhere; was that combination intentional?

This looks like it was a mistake and can just be removed. The Xdock arguments no longer do anything. 

I took a look at the changes so far and I have a few high level comments:

For these two files, could you move these changes into a separate PR/branch? They're CI infrastructure changes and not related to what we ship.

- `.github/hub_scripts/pr_ci.sh`
- `subprojects/performance/docs/check-rev.sh`

`.teamcity/mvnw` is also a CI infrastructure change, but I think it may be auto-generated, so it's likely your changes will get overwritten at some point. The upstream project for this is Apache Maven, so I think it's OK if you just drop these changes for now.

You don't need to worry about making changes directly to `gradlew` since it's generated from `unixStartScript.txt`. What will happen is your changes will get incorporated into a nightly version of Gradle and when we update to use that nightly the `gradlew` script will be regenerated.

So I think that means you only need to submit changes for `unixStartScript.txt`. If you run `./gradlew plugins:quickTest` locally that should test enough to shake out a big breakage. I can trigger a full pipeline that will test macOS/Windows/Linux more thoroughly.

**Comment by kurahaupo** [comment]
I'm working on unwinding the changes to `.github/hub_scripts/pr_ci.sh`, `subprojects/performance/docs/check-rev.sh`, and `.teamcity/mvnw`.

When other projects use the template, what do they call the wrapper script? Is it always `gradlew`?
If not, is there a template substitution parameter that contains the name of the script, that I can use in the documentation block at the top?

File: subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt (hunk 3 above)
Comment (kurahaupo) [reply subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt:81]: That was my unfamiliarity with Groovy; the comment needs `/*...*/` rather than `#`, so I'm adjusting that.

**Comment by kurahaupo** [comment]
@big-guy I do appreciate your testing, as I'm currently stuck with

    $ ./gradlew plugins:quickTest
    Gradle 7.2-20210704221017+0000 requires Java 8 or later to run. You are currently using Java 7.

I understand that as `gradlew` is a build artefact it doesn't technically need to be patched, but my inability to do local testing is one reason that I've left these patches in place, so that if there are any questions about specific commits, or any changes requested, I can modify `gradlew` and then use _that_ as the basis for patching the template. Let me know if you need these stripped out, or whether you're OK to simply let them be overwritten by the next nightly build after merging.

Obviously I need to update Java before I can _use_ Gradle, and that's been my work in progress for a while, but when I noticed the recent updates I thought I should send my PR before things diverge any further.

I'm also rebasing my changing onto `master` before each submission to ensure a clean merge. What's your policy on doing that?

File: subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt (hunk 1 above)
Comment (kurahaupo) [reply subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt:21]: Done

File: subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt (hunk 2 above)
Comment (kurahaupo) [reply subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt]: I've re-organized this.

My objective is to provide a starting point for someone like myself, who's a seasoned Unix shell user, but who hasn't done Java programming in a modern framework. (The last Java program I wrote targetted Java version 1.4, and everything was done on the command line, and I had a Makefile for building jars.)

When the instructions say _run_ this script, the first thing I do is _read_ the script.

So I'm including what to you must seem like a great deal of redundant information, but it's not redundant to the target audience.

**Comment by kurahaupo** [comment]
Of course, the other audience is to explain some fairly nuanced shell script to people whose primary skillset is Java. I hope my comments succeed in that respect; I would certainly appreciate feedback if anything is unclear.

**Comment by kurahaupo** [comment]
Sorry, pushed again as I forgot to elide the change to `check-rev.sh`.

**Comment by kurahaupo** [comment]
@big-guy github still thinks I have outstanding changes requested by you but I can't find them (probably because they're based on commits that I've rebased out of existence).

If you have time to take another look that would be appreciated.

**Comment by kurahaupo** [comment]
Do you need me to write any additional tests, seeing as how there's "in theory" no modified behaviour.

I can think of some cases where behaviour might change:

1. Under CygWin and Msys, `$GRADLE_CYGPATTERN` is optionally used to detect when a string is a filename that should be rewritten as a Windows path. Because I've changed the order of tests, it's theoretically possible that it now won't match a valid filename starting with a `-`, however I don't see how that could be useful because `cygpath` would have interpreted such as string as an option, not a filename.

This also raises the possibility of _bugs I should fix while I'm at it_:
1. Also for Cygwin & Msys, should the expansions of `$DEFAULT_JVM_OPTS`, `$JAVA_OPTS`, and `$GRADLE_OPTS` be subject to file path cygpath rewriting? (They weren't, and they're still not, but should they be?)
2. The `die` and `warn` functions [lines 119-128](/kurahaupo/gradle/blob/93f7f594e768c175fc9e99524902b6d992c7f1ad/gradlew#L119) should probably write to stderr rather than stdout.
3. The command `ulimit -H -n` [line 169](/kurahaupo/gradle/blob/93f7f594e768c175fc9e99524902b6d992c7f1ad/gradlew#L169) is used to compute a value which may wind up not being used; in that case there's no point issuing a warning [line 177](/kurahaupo/gradle/blob/93f7f594e768c175fc9e99524902b6d992c7f1ad/gradlew#L177).

File: subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt [hunk 4]
```diff
@@ -17,37 +17,145 @@
...
+#       within the Gradle project, so that's where you need to submit any
+#       patches.
+#
+#       Gradle is a meta-build system used by the project that you're building
+#       or installing. It's like autoconf but for projects that are written in
+#       Java and related languages. It's also used to build parts of the Gradle
+#       project itself.
+#
+#       Gradle is mostly written in Java and Groovy, but there are wrapper
+#       scripts for various environments, including this one for POSIX shells.
```
Comment (big-guy) [inline subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt:56]: Since this is something that gets copied into a lot of projects, we need to be considerate with the amount of detail here.

I think the comments that are left in this file should describe how this script works and where this script comes from.

e.g., I was thinking of summarizing your (1)-(3) with:
```
#
#   (1) You need a POSIX-compliant shell to run this script. If your /bin/sh is
#       noncompliant, but you have some other compliant shell such as ksh or
#       bash, then to run this script, type that shell name before the whole
#       command line, like:
#
#           ksh $applicationName
# 
#   (2) This script targets any POSIX shell, not just Bash or Ksh.
#
#       It makes use of these POSIX shell features:
#         * functions;
#         * various expansions, including «$var», «$( cmd )», «${var#prefix}»,
#           and «${var%suffix}»;
#         * compound commands having a testable exit status, especially «case»;
#         * various built-in commands including «command», «set», and «ulimit».
#
#       However arrays are NOT available, except for "\$@".
#
#       This script progressively modifies "\$@", which is eventually
#       passed to Java.
#
#       There is special handling for the inherited environment variables
#       DEFAULT_JVM_OPTS, JAVA_OPTS, and ${optsEnvironmentVar}; see the in-line
#       comments for details.
#
#       There are tweaks for specific operating systems such as AIX, CygWin,
#       Darwin, MinGW, and NonStop.
#
#   (3) This script is generated from a Groovy template file in Gradle
#       https://github.com/gradle/gradle/blob/master/subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt
#
```

File: subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt [hunk 5]
```diff
@@ -17,37 +17,145 @@
...
+#           ${appNameSystemProperty}    org.gradle.appname
+#           ${applicationName}          Gradle
+#           ${mainClassName}            org.gradle.wrapper.GradleWrapperMain
+#           ${optsEnvironmentVar}       GRADLE_OPTS
+#           ${defaultJvmOpts}           -Dfile.encoding=UTF-8 "-Xmx64m" "-Xms64m"
+#
+#       (These are all specified in
+#       subprojects/build-init/src/main/java/org/gradle/api/tasks/wrapper/Wrapper.java
+#       within the Gradle project; they may be different for other projects.)
+# */ %>
```
Comment (big-guy) [inline subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt:97]: I would take some variation of what you have here between <%/%> and move it to subprojects/plugins/src/main/java/org/gradle/api/internal/plugins/UnixStartScriptGenerator.java

That's the class that uses this template to generate the script. If you think it would be helpful for future readers, we could add a link back to UnixStartScriptGenerator from the template that is not included in the generated script. My main concern with packing too much developer documentation in the template itself is that it's already confusing and unnatural to escape everything.

File: subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt [hunk 6]
```diff
@@ -17,37 +17,145 @@
...
+#       The is special handling for the inherited environment variables
+#       DEFAULT_JVM_OPTS, JAVA_OPTS, and ${optsEnvironmentVar}; see the in-line
+#       comments for details.
+#
+#       There are tweaks for specific operating systems such as AIX, CygWin,
+#       Darwin, MinGW, and NonStop.
+#
+#       If submitting patches, please make sure that this script (as output by
+#       the template generator) still passes shellcheck
+#       (https://shellcheck.net)
```
Comment (big-guy) [inline subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt:127]: I think we can enforce this on our CI infrastructure as well now that the script passes. You don't need to figure that out.

**Review by big-guy — CHANGES_REQUESTED** [review]
Hopefully I haven't missed any of your questions.

> When other projects use the template, what do they call the wrapper script? Is it always `gradlew`?
> If not, is there a template substitution parameter that contains the name of the script, that I can use in the documentation block at the top?

It's based on the name of the application. I believe it's the same thing as `applicationName`.

> I understand that as gradlew is a build artefact it doesn't technically need to be patched, but my inability to do local testing is one reason that I've left these patches in place, so that if there are any questions about specific commits, or any changes requested, I can modify gradlew and then use that as the basis for patching the template. Let me know if you need these stripped out, or whether you're OK to simply let them be overwritten by the next nightly build after merging.

Probably what I'll do is squash your changes into another branch and remove those changes so other people don't get confused why `gradlew` changed without a nightly version bump. So it's OK for you to keep these changes if it's helping you test locally.

> Also, I'm rebasing my changing onto master before each submission to ensure a clean merge. What's your policy on doing that?

For the most part, you don't need to do that unless we ask. Sometimes we make CI changes that require old branches/PRs to be rebased, but we have a merge queue that does the merge to master. For where you're making changes, it's very unlikely anyone else is going to make changes in that area right now. 

> Do you need me to write any additional tests, seeing as how there's "in theory" no modified behaviour.

I think we can rely on our existing coverage plus the RC process to identify problems. I've already run an earlier version of your changes through our CI pipeline and I didn't find anything surprising. 

> I can think of some cases where behaviour might change: <snip>

I think if the behavior is different as you describe, it will be OK as well. I'll double check that we're running some of our tests through a cygwin shell on Windows.

> Also for Cygwin & Msys, should the expansions of $DEFAULT_JVM_OPTS, $JAVA_OPTS, and $GRADLE_OPTS be subject to file path cygpath rewriting? (They weren't, and they're still not, but should they be?)

No, let's leave this as-is. It would be unusual for any of those to include file paths without it being a part of a larger argument. e.g., `-DFOO=/path/to/foo` would be hard to parse out and cygpath rewrite. 

> The die and warn functions (in lines 119-128) should probably write to stderr rather than stdout.

+1, this is a good idea for die. I don't think there's any good reason it's not doing that now. warn is OK on stdout.

> The command ulimit -H -n at line 169 is used to compute a value which may wind up not being used; in that case there's no point issuing a warning at line 177.

+1, I think ditching the warning makes sense to me. 

> Should Darwin really be excluded from setting ulimit -n? Darwin has a BSD UNIX kernel, so it should just work.

From the history, I can't immediately tell if this was intentional or not. Let's leave this as-is and get the rest of the improvements you made into master first.

File: subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt (hunk 4 above)
Comment (kurahaupo) [reply subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt:56]: That looks like a reasonable balance.
Enough detail so that the reader of `gradlew` can understand it (and why it is written the way that it is), and a link so somewhere with more detail.

File: subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt (hunk 5 above)
Comment (kurahaupo) [reply subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt:97]: I take your point on size, but this is only visible to someone reading the template, at which point it's critically important that they understand the three-phase parsing so I'd rather leave that part in place.

On the other hand, they probably don't need the list of available Groovy variables unless they're making substantial changes, in which case a link elsewhere would suffice.

I assume some kind of JavaDoc format would be appropriate? I'm very rusty on what that would entail. Perhaps that should be a separate PR?

File: subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt [hunk 7]
```diff
@@ -17,37 +17,145 @@
 #
 
 ##############################################################################
-##
-##  ${applicationName} start up script for UN*X
-##
+#
+#   ${applicationName} start up script for POSIX<% if ( applicationName != 'Gradle' ) { %>
+#   generated by Gradle<% } %>.
```
[the remark this answers is not in the document]
Comment (kurahaupo) [reply subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt:22]: I now notice some of the test suite seems to look for "UN*X"; should I change them to "POSIX" as well?

**Comment by kurahaupo** [comment]
(This explains a change that will be squashed when I next push my branch.)

Prompted by [your use of quotes on line 73](https://github.com/gradle/gradle/blob/sg/72/unixScript/subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt#L43) I looked more closely at all the quotes and assignments.

The short version is that I found a few places where I'd messed up by removing quotes (because I didn't account for templating), but if I simply put them back the result is still problematic: Groovy can in principle supply variables that contain quote marks and other shell metacharacters.

So I'm working on a new method that looks like

```
read -r variable <<EndOfGroovySubstitution
$groovyVar
EndOfGroovySubstitution
```

This means that it's possible for `$groovyVar` to expand to `$SHELL_VAR` and have that expanded in turn, but without risking blowing up the shell if there are quote marks in `$groovyVar`. (Such quote marks might not lead to the desired outcome of course, but the error reporting will at least make sense.)

This doesn't apply in all cases; java class names consist of alphanumerics and dots, so they're safe for the shell, and the odd case where 'mainClass' consists of `--module` and a Java class name is also safe. However perhaps this should be documented.


I'm also looking at using a similar technique to remove the `sed`+`eval` pairing, which for most script writers will be a black box.

The naïve approach would be to write something like this:
```
echo " $SHELL_VAR " |
xargs -n1 |
while read -r line ; do
    set -- "$@" "$line"
done
```
however that won't in all POSIX shells, and in particular won't work in Bash without prefacing it with `shopt -s lastpipe` which itself is not portable.

If one has Bash then one can use process substitution:
```
while read -r line ; do
    set -- "$@" "$line"
done < <(
    echo " $SHELL_VAR " |
    xargs -n1
)
```
but that's still not POSIX.

I think this is the simplest version that will actually work for all POSIX shells:
```
while read -r line ; do
    set -- "$@" "$line"
done <<EndOfExpansion
$(
    echo " $SHELL_VAR " |
    xargs -n1
)
EndOfExpansion
```

There is also the option of a pure-shell version, that does not rely on `xargs` or `sed` at all, but it's not for the novice shell scripter:
```
unset next in_quote
while [ -n "$SHELL_VAR" ] ; do
    t=${SHELL_VAR#?}
    h=${SHELL_VAR%"$t"}
    SHELL_VAR=$t
    case $h in
      $in_quote)
        unset in_quote
        continue
        ;;
      ${in_quote-[[:space:]]})
        [ -n "${next+set}" ] && set -- "$@" "$next"
        unset next
        ;;
      [\"\'])
        in_quote=$h
        ;;
      \\)
        t=${SHELL_VAR#?}
        h=${SHELL_VAR%"$t"}
        SHELL_VAR=$t
        next=$next$h
        ;;
      *)
        next=$next$h
        ;;
    esac        
done
[ -n "${next+set}" ] && set -- "$@" "$next"

```

**Comment by kurahaupo** [comment]
In addition to revamping the assignments above, I'm also looking at using a related technique to remove this `sed`+`eval` pairing:

```
eval "set -- $(
        printf '%s\n' "$DEFAULT_JVM_OPTS $JAVA_OPTS $GRADLE_OPTS" |
        xargs -n1 |
        sed ' s~[^-[:alnum:]+,./:=@_]~\\&~g; ' |
        tr '\n' ' '
    )" '"$@"'
```

While it's a bit simpler than it was, I'm sure it's still a mysterious black box for most script writers.

The naïve approach to replacing it would be to write something like this:
```
echo "$DEFAULT_JVM_OPTS $JAVA_OPTS $GRADLE_OPTS" |
xargs -n1 |
while read -r line ; do
    set -- "$@" "$line"
done
```
however that won't work in all POSIX shells, because POSIX allows the shell to run _every_ part of a pipeline in a separate subshell (except for the degenerate case of a "pipeline" that has only a single part). In particular won't work in Bash without prefacing it with `shopt -s lastpipe` (which itself is not portable, and is only available in newer versions of Bash).

If one has Bash then one can use process substitution:
```
while read -r line ; do
    set -- "$@" "$line"
done < <(
    echo "$DEFAULT_JVM_OPTS $JAVA_OPTS $GRADLE_OPTS" |
    xargs -n1
)
```
but that's still not POSIX.

I think this is the simplest version that will actually work for all POSIX shells:
```
while read -r line ; do
    set -- "$@" "$line"
done <<EndOfExpansion
$(
    echo "$DEFAULT_JVM_OPTS $JAVA_OPTS $GRADLE_OPTS" |
    xargs -n1
)
EndOfExpansion
```

There is also the option of a pure-shell version, that does not rely on `xargs` or `sed` at all, but it's not for the novice shell scripter:
```
# NB: make sure that word_list ends with whitespace, so that we will gather the last word into "$@"
word_list="$DEFAULT_JVM_OPTS $JAVA_OPTS $GRADLE_OPTS "
unset next in_quote   # initial Lexer state:
while [ -n "$word_list" ] ; do
    # pop one character off the front of word_list
    t=${word_list#?}
    h=${word_list%"$t"}
    word_list=$t
    # decide what to do with it
    case $h in
      $in_quote)
        # we were in quotes, so we've just found the close quote
        unset in_quote
        ;;
      ${in_quote-[[:space:]]})
        # we were not in quotes AND the char is whitespace, so it's a word break
        [ -n "${next+set}" ] && set -- "$@" "$next"
        unset next
        ;;
      \\)
        # backslash escapes the next character:
        # pop the next characer off the front of SHELL_VAR (same as above) ...
        t=${word_list#?}
        h=${word_list%"$t"}
        word_list=$t
        # ... and add it to the current word (same as any ordinary character, below)
        next=$next$h
        ;;
      [\"\'])
        # we weren't in quotes, so we are now
        in_quote=$h
        ;;
      *)
        # any ordinary character gets added to the current word (or starts a new word)
        next=$next$h
        ;;
    esac        
done
```

**Comment by kurahaupo** [comment]
(in reference to the script name)
> It's based on the name of the application. I believe it's the same thing as `applicationName`.

The script name is `gradlew` while `$applicationName` expands to `Gradle`. Perhaps something like `${applicationName.toLower()}w` might work, but it feels really kludgy and I'm trying to get rid of kludges.

I'll just try to get rid of named references to the script within itself.

**Comment by kurahaupo** [comment]
> Probably what I'll do is squash your changes into another branch and remove those changes so other people don't get confused why `gradlew` changed without a nightly version bump. So it's OK for you to keep these changes if it's helping you test locally.

I'll do a rebase and strip them myself once we're agree that it's all satisfactory; that way I get to keep the attribution. And it will keep the changes segmented in separate commits for anyone who looks back and wonders "why".

**Comment by big-guy** [comment]
> Groovy can in principle supply variables that contain quote marks and other shell metacharacters.

I think this is OK to ignore for the changes you're doing now. This was likely already a problem and there will be other places where assumptions have been made about the variables substituted in the script. We could also move the escaping up into the template expansion (in Java code) to handle this if that would make the shell script simpler.

**Comment by big-guy** [comment]
@kurahaupo Just to keep you in the loop on things.  

First off, thanks for the contribution. I think you've really improved what we had and being able to keep POSIX compatibility is a big plus. 

We're trying to prepare for the next release (7.2), so I merged this PR into a separate branch and reworked a couple of things: https://github.com/gradle/gradle/pull/17716

1. Like I said above, I just reverted the changes to `gradlew`. 
2. I also added some additional test coverage for questions that I had with some of the changes.
3. I removed the use of `SHELL_VAR=\$( cat ) <<EndOfGroovySubstitution` for substitutions. This seemed to be the source of some test failures, so I checked that we escape all/most of shell metacharacters before substituting them. It's unlikely that these substitutions would actually contain metacharacters anyways. 
4. I also tried to make some of our existing tests less sensitive to the exact implementation of the script (e.g., we used to check for an exact number of lines).

I've built a new nightly with your changes and I've submitted another PR to bump the version we use on gradle/gradle: 
https://github.com/gradle/gradle/pull/17733 

If you'd like to try out the latest changes, you can use upgrade to this nightly:
```
./gradlew wrapper --gradle-version=7.2-20210715231856+0000 && ./gradlew wrapper
```

The only thing that I found that seems to be broken has to do with unbalanced quotes. It seems like you used to be able to pass these through environment variables, but you can't do that any longer. I can't really think of a good reason you'd want to do that and I think the benefits of all the other work you did outweigh this breakage (if it is one). I'm going to include a note about this in our upgrade guide, but I don't think we need to fix it without a good example.

If there's more that you'd like to improve, please feel free to submit another PR with those changes and ping me.

**Comment by kurahaupo** [comment]
yeah I noticed that some of the tests look for explicit lines in the generated files, so I adjusted them but perhaps I missed something.

The issue is not just unbalanced quotes; it's that different quotes are required/allowed inside different variables; whichever is not used to enclose the assignment in the template.

You can get away with using the same quote type as long as they are balanced. If the groovy substitution result in say
 var=""foo""
then that is valid but _foo_ is UNquoted, so if it contains other shell meta-characters (space, tab, newline, and ;&()|\<>) it will break.

If single quotes are used in the template to prevent $ and ` from triggering expansion, then having single quotes in the value inserted by the template engine will defeat that, leaving them subject to expansion.

Really this should be fixed by having groovy insert \ before each shell meta-character, but that might still need adjustments to the test suite.

That in turn means that Groovy should be made aware of what is or isn't an array value, which on balance is a good thing, but would be a user-visible change.

However you're right that these can be improved as a separate PR.

I've managed to get a local build running so I will drop the latest patch and all the patches to gradlew, and push the branch again when I get home tonight.
